File: sdk/python/ScyKernel.py
import os
import mmap
import math
import zlib
import struct
class ScyKernel:
    def __init__(self, password, file_path):
        self.password = password
        self.file_path = file_path
        self.canvas_size = 4000
        self.h_val = self.get_h_val(password)
        self.db_buffer = bytearray()
    def get_h_val(self, pwd):
        hash_val = 7
        for char in pwd:
            hash_val = (hash_val * 31 + ord(char)) & 0xFFFFFFFF
        normalized = hash_val / 4294967296.0
        return int(math.floor(normalized * 16000000.0))
    def derive_index(self, key, password):
        hash_val = 0x811c9dc5
        prime = 0x01000193
        alpha_salt = 0
        if password:
            for char in password:
                hash_val = (hash_val ^ ord(char)) & 0xFFFFFFFF
                hash_val = (hash_val * prime) & 0xFFFFFFFF
        for char in key:
            b = ord(char)
            hash_val = (hash_val ^ b) & 0xFFFFFFFF
            hash_val = (hash_val * prime) & 0xFFFFFFFF
            if char.isalpha():
                alpha_salt += (ord(char.lower()) - ord('a') + 1)
        final_val = (hash_val + alpha_salt) & 0xFFFFFFFF
        normalized = final_val / 4294967296.0
        return int(math.floor(normalized * 16000000.0))
    def d2xy(self, n, d):
        x = y = 0
        t = d
        s = 1
        while s < n:
            rx = 1 & (t // 2)
            ry = 1 & (t ^ rx)
            x, y = self.rot(s, x, y, rx, ry)
            x += s * rx
            y += s * ry
            t //= 4
            s *= 2
        return x, y
    def rot(self, n, x, y, rx, ry):
        if ry == 0:
            if rx == 1:
                x = n - 1 - x
                y = n - 1 - y
            return y, x
        return x, y
    # A bit-perfect, zero-byte overhead encryption layer.
    # XORs the character based on the password's hash and the character's position.
    def crypt_byte(self, c, password, position):
        salt = 0x811c9dc5
        for pc in password:
            salt = ((salt ^ ord(pc)) * 16777619) & 0xFFFFFFFF
        mixed = (salt ^ (position * 0xdeadbeef)) & 0xFFFFFFFF
        mixed ^= (mixed >> 16)
        key_byte = mixed & 0xFF
        return chr(ord(c) ^ key_byte)
    # Crypt byte sanitizer to prevent null collisions
    def sanitize_and_xor(self, raw_char, reverse=False):
        a = 123
        c = 45
        val = ord(raw_char)
        if reverse:
            val_adjusted = val - 1
            inv_a = 179
            return chr(((val_adjusted - c) * inv_a) % 256)
        else:
            transformed = (val * a + c) % 256
            return chr(transformed + 1)
    def sanitize_and_xor(self, raw_char, reverse=False):
        return raw_char
    # Lightweight CRC-32 for PNG Chunk Compliance
    def compute_crc(self, buf):
        crc = 0xFFFFFFFF
        for b in buf:
            crc ^= b
            for _ in range(8):
                mask = -(crc & 1) & 0xFFFFFFFF
                crc = (crc >> 1) ^ (0xEDB88320 & mask)
        return (~crc) & 0xFFFFFFFF
    def write32(self, out, val):
        out.write(struct.pack(">I", val & 0xFFFFFFFF))
    # PPM sow, harvest functions
    def put_to_ppm(self,key,value,password):
        index=self.derive_index(key,password)
        cur_d=self.h_val+(index*1600)
        x,y=self.d2xy(self.canvas_size,cur_d)
        try:
            with open(self.file_path,"r+b") as file:
                offset=15+(y*self.canvas_size+x)*3
                file.seek(offset)
                for i,c in enumerate(value):
                    raw_data = file.read(3)
                    if len(raw_data) < 3:
                        break
                    pixel=bytearray(raw_data)
                    secure_char=self.crypt_byte(c,password,i)
                    sanitized_char = self.sanitize_and_xor(secure_char, reverse=False)
                    pixel[0]=ord(sanitized_char)&0xFF
                    file.seek(-3,1)
                    file.write(pixel)
                    file.seek(file.tell())
                file.write(bytes([0,0,0]))
        except IOError:return
    def get_from_ppm(self, key, password):
        index = self.derive_index(key, password)
        cur_d = self.h_val + (index * 1600)
        x, y = self.d2xy(self.canvas_size, cur_d)
        terminator_count = 0
        try:
            with open(self.file_path, "rb") as file:
                offset = 15 + (y * self.canvas_size + x) * 3
                file.seek(offset)
                result = []
                i = 0
                while True:
                    pixel = file.read(3)
                    if not pixel: break
                    if pixel == b"\x00\x00\x00":
                        terminator_count += 1
                    else:
                        terminator_count = 0
                    restored_char = self.sanitize_and_xor(chr(pixel[0]), reverse=True)
                    result.append(self.crypt_byte(restored_char, password, i))
                    if terminator_count >= 5:
                        #print(f"PPM Termination occurred at pixel {i}.")
                        result = result[:-5]
                        break
                    i += 1
                return "".join(result)
        except IOError:
            return ""
    # PNG sow, harvest functions
    # Writes a value to the RAM buffer.
    # Note: You MUST call sync_png(file, "commit") after calling this to save changes.
    def put_to_png(self, key, value, key_password):
        if not self.db_buffer:
            self.db_buffer = bytearray(48000000)
        index = self.derive_index(key, key_password)
        cur_d = self.h_val + (index * 1600)
        x, y = self.d2xy(self.canvas_size, cur_d)
        for i in range(len(value)):
            pixel_idx = ((y * self.canvas_size) + (x + i)) * 3
            if pixel_idx + 2 < len(self.db_buffer):
                secure_char = self.crypt_byte(value[i], key_password, i)
                sanitized_char = self.sanitize_and_xor(secure_char, reverse=False)
                self.db_buffer[pixel_idx] = ord(sanitized_char) & 0xFF
        term_idx = ((y * self.canvas_size) + (x + len(value))) * 3
        if term_idx + 2 < len(self.db_buffer):
            self.db_buffer[term_idx] = 0
    # Retrieves a value from the RAM buffer.
    # Note: You SHOULD call sync_png(file, "load") before this to ensure fresh data.
    def get_from_png(self, key, key_password):
        if not self.db_buffer:
            return ""
        index = self.derive_index(key, key_password)
        cur_d = self.h_val + (index * 1600)
        x, y = self.d2xy(self.canvas_size, cur_d)
        result = []
        terminator_count = 0
        i = 0
        while True:
            pixel_idx = ((y * self.canvas_size) + (x + i)) * 3
            if pixel_idx + 2 >= len(self.db_buffer):
                break
            r, g, b = self.db_buffer[pixel_idx : pixel_idx + 3]
            if r == 0 and g == 0 and b == 0:
                terminator_count += 1
            else:
                terminator_count = 0
            restored_char = self.sanitize_and_xor(chr(self.db_buffer[pixel_idx]), reverse=True)
            result.append(self.crypt_byte(restored_char, key_password, i))
            if terminator_count >= 5:
                #print(f"PNG Termination occurred at pixel {i}.")
                result = result[:-5]
                break
            i += 1
        return "".join(result)
    # DB sync functions for easier DB handling
    def sync_png(self,filename,mode):
        mode=mode.lower()
        if mode=="load" and not os.path.exists(filename):
            self.db_buffer=bytearray(48000000)
            return self.sync_png(filename,"commit")
        if mode=="commit":
            filtered=bytearray()
            for r in range(4000):
                filtered.append(0)
                start=r*12000
                filtered.extend(self.db_buffer[start:start+12000])
            compressed=zlib.compress(filtered)
            with open(filename,"wb") as out:
                out.write(bytes([0x89,0x50,0x4E,0x47,0x0D,0x0A,0x1A,0x0A]))
                ihdr=bytes([73,72,68,82,0,0,15,160,0,0,15,160,8,2,0,0,0])
                self.write32(out,13)
                out.write(ihdr)
                self.write32(out,self.compute_crc(ihdr))
                self.write32(out,len(compressed))
                out.write(b"IDAT")
                out.write(compressed)
                crc_b=b"IDAT"+compressed
                self.write32(out,self.compute_crc(crc_b))
                self.write32(out,0)
                out.write(b"IEND")
                self.write32(out,self.compute_crc(b"IEND"))
            return True
        elif mode=="load":
            if os.path.getsize(filename)<37:return False
            with open(filename,"rb") as file:
                file.seek(33)
                raw_len=file.read(4)
                if len(raw_len)<4:return False
                c_len=struct.unpack(">I",raw_len)[0]
                file.seek(4,1)
                c_data=file.read(c_len)
            try:
                decomp=zlib.decompress(c_data)
                self.db_buffer=bytearray(48000000)
                for r in range(4000):
                    start=r*12001+1
                    self.db_buffer[r*12000:(r+1)*12000]=decomp[start:start+12000]
                return True
            except:return False
        else:return False
    def sync_ppm(self, filename, mode):
        mode = mode.lower()
        if mode == "load" and not os.path.exists(filename):
            self.db_buffer = bytearray(48000000)
            return self.sync_ppm(filename, "commit")
        if mode == "commit":
            with open(filename, "wb") as out:
                out.write(b"P6\n4000 4000\n255\n")
                out.write(self.db_buffer)
            return True
        elif mode == "load":
            with open(filename, "rb") as file:
                file.seek(15)
                self.db_buffer = bytearray(file.read(48000000))
            return True
        else:return False
    # Blank DB creation functions
    def create_png_db(self, filename):
        self.db_buffer = bytearray(48000000)
        self.sync_png(filename, "commit")
    def create_ppm_db(self, db_path):
        try:
            with open(db_path, "wb") as ofs:
                header = f"P6\n{self.canvas_size} {self.canvas_size}\n255\n".encode()
                ofs.write(header)
                zero_row = bytearray(self.canvas_size * 3)
                for _ in range(self.canvas_size):
                    ofs.write(zero_row)
        except IOError:return
    # DB conversion functions
    def convert_database_format(self, png_path, ppm_path, target_format):
        target_format = target_format.lower()
        if target_format == "ppm":
            if not self.sync_png(png_path, "load"):return False
            with open(ppm_path, "wb") as ppm:
                ppm.write(b"P6\n4000 4000\n255\n")
                ppm.write(bytes(self.db_buffer))
            return True
        elif target_format == "png":
            with open(ppm_path, "rb") as ppm:
                ppm.seek(15)
                self.db_buffer = bytearray(ppm.read(48000000))
            if not self.sync_png(png_path, "commit"):return False
            return True
        else:return False
    # DB Deletion and cleanup functions
    def delete_db(self, db_path: str) -> bool:
        """Removes the database file if it exists."""
        try:
            if os.path.exists(db_path):
                os.remove(db_path)
                return True
            return False
        except OSError:return False
    # DB checking/validating functions
    def get_file_size(self, path):
        try:
            if os.path.exists(path) and os.path.isfile(path):
                return os.path.getsize(path)
        except OSError:return 0
        return 0
    def validate_db(self, path):
        raw_data_size = 48000000
        actual = self.get_file_size(path)
        if ".ppm" in path:
            return actual >= (raw_data_size + 15)
        else:
            return actual >= 1000

File: sdk/javascript/ScyKernel.js
const fs = require('fs');
const zlib = require('zlib');

class ScyKernel {
    constructor(password, filePath) {
        this.password = password;
        this.filePath = filePath;
        this.canvasSize = 4000;
        this.hVal = this._getHVal(password);
        this.dbBuffer = null;
    }

    _getHVal(pwd) {
        let hash;
        hash = 7;
        for (let i = 0; i < pwd.length; i++) {
            hash = (Math.imul(hash, 31) + pwd.charCodeAt(i)) | 0;
        }
        let unsignedHash;
        unsignedHash = hash >>> 0;
        let normalized;
        normalized = unsignedHash / 4294967296.0;
        let result;
        result = Math.floor(normalized * 16000000.0);
        return result;
    }

    // Deterministic FNV-1a + Alphabet Salt for Parity with option
    // to support password protected databasing
    _deriveIndex(key, password) {
        let hash;
        hash = 0x811c9dc5 >>> 0;
        let prime;
        prime = 0x01000193;
        let alphaSalt;
        alphaSalt = 0;
        if (password && password.length > 0) {
            for (let i = 0; i < password.length; i++) {
                let pChar;
                pChar = password.charCodeAt(i) & 0xFF;
                hash = (hash ^ pChar) >>> 0;
                hash = Math.imul(hash, prime) >>> 0;
            }
        }
        for (let i = 0; i < key.length; i++) {
            let kCharCode;
            kCharCode = key.charCodeAt(i);
            hash = (hash ^ (kCharCode & 0xFF)) >>> 0;
            hash = Math.imul(hash, prime) >>> 0;
            let c;
            c = key[i];
            if (/[a-zA-Z]/.test(c)) {
                alphaSalt += (c.toLowerCase().charCodeAt(0) - 97 + 1);
            }
        }
        let finalVal;
        finalVal = (hash + alphaSalt) >>> 0;
        let normalized;
        normalized = finalVal / 4294967296.0;
        let result;
        result = Math.floor(normalized * 16000000.0);
        return result;
    }

    _d2xy(n, d) {
        let rx;
        let ry;
        let s;
        let t;
        t = d;
        let x;
        x = 0;
        let y;
        y = 0;
        for (s = 1; s < n; s *= 2) {
            rx = 1 & (Math.floor(t / 2));
            ry = 1 & (t ^ rx);
            let rotated;
            rotated = this._rot(s, x, y, rx, ry);
            x = rotated.x;
            y = rotated.y;
            x = x + (s * rx);
            y = y + (s * ry);
            t = Math.floor(t / 4);
        }
        return { x, y };
    }
    
    _rot(n, x, y, rx, ry) {
        let targetX;
        targetX = x;
        let targetY;
        targetY = y;
        if (ry === 0) {
            if (rx === 1) {
                targetX = n - 1 - targetX;
                targetY = n - 1 - targetY;
            }
            let temp;
            temp = targetX;
            targetX = targetY;
            targetY = temp;
        }
        return { x: targetX, y: targetY };
    }

    /**
     * A bit-perfect, zero-byte overhead encryption layer.
     * XORs the character based on the password's hash and the character's position.
     */
    _cryptByte(c, password, position) {
        let salt;
        salt = 0x811c9dc5 >>> 0;
        let prime;
        prime = 16777619;
        for (let i = 0; i < password.length; i++) {
            let charByte;
            charByte = password.charCodeAt(i) & 0xFF;
            salt = Math.imul(salt ^ charByte, prime) >>> 0;
        }
        let posMult;
        posMult = Math.imul(position, 0xdeadbeef) >>> 0;
        let mixed;
        mixed = (salt ^ posMult) >>> 0;
        mixed ^= (mixed >>> 16);
        let keyByte;
        keyByte = mixed & 0xFF;
        let result;
        result = (c & 0xFF) ^ (keyByte & 0xFF);
        return result & 0xFF;
    }

    // Lightweight CRC-32 for PNG Chunk Compliance
    _computeCRC(buf, len) {
        let crc = 0xffffffff >>> 0;
        for (let i = 0; i < len; i++) {
            crc ^= buf[i];
            for (let j = 0; j < 8; j++) {
                // Using -(crc & 1) logic via bitwise mask
                if (crc & 1) {
                    crc = (crc >>> 1) ^ 0xedb88320;
                } else {
                    crc = crc >>> 1;
                }
            }
        }
        return (crc ^ 0xffffffff) >>> 0;
    }

    _write32(val) {
        const buf = Buffer.alloc(4);
        buf.writeUInt32BE(val, 0);
        return buf;
    }

    // PPM sow, harvest functions
    _putToPPM(key, value, password) {
        const index = this._deriveIndex(key, password);
        const curD = this.hVal + (index * 1600);
        const { x, y } = this._d2xy(this.canvasSize, curD);
        try {
            const fd = fs.openSync(this.filePath, 'r+');
            // Header offset (Standard P6 PPM is 15 bytes)
            let offset = 15 + (y * this.canvasSize + x) * 3;
            for (let i = 0; i < value.length; i++) {
                const pixel = Buffer.alloc(3);
                // Read 3 bytes (the current pixel)
                fs.readSync(fd, pixel, 0, 3, offset);
                // Apply ScyKernel stream cipher
                const charCode = value.charCodeAt(i);
                const secureByte = this._cryptByte(charCode, password, i);
                // Update Red channel
                pixel[0] = secureByte;
                // Overwrite the same pixel
                fs.writeSync(fd, pixel, 0, 3, offset);
                // Advance offset by 3 bytes (1 pixel)
                offset += 3;
            }
            // Write Null Terminator (Red channel = 0)
            const term = Buffer.from([0, 0, 0]);
            fs.writeSync(fd, term, 0, 3, offset);
            fs.closeSync(fd);
        } catch (err) {
            // console.error(`❌ PPM Write Error: ${err.message}`);
        }
    }

    _getFromPPM(key, password) {
        const index = this._deriveIndex(key, password);
        const curD = this.hVal + (index * 1600);
        const { x, y } = this._d2xy(this.canvasSize, curD);
        try {
            const fd = fs.openSync(this.filePath, 'r');
            let offset = 15 + (y * this.canvasSize + x) * 3;
            let result = "";
            let i = 0;
            while (true) {
                const pixel = Buffer.alloc(3);
                const bytesRead = fs.readSync(fd, pixel, 0, 3, offset);
                // Bounds check: Stop if EOF or Red channel is 0 (Null Terminator)
                if (bytesRead < 3 || pixel[0] === 0) {
                    break;
                }
                const scrambled = pixel[0];
                const decryptedByte = this._cryptByte(scrambled, password, i);
                result += String.fromCharCode(decryptedByte);
                offset += 3;
                i++;
            }
            fs.closeSync(fd);
            return result;
        } catch (err) {
            return "";
        }
    }

    // PNG sow, harvest functions
    /**
    * Writes a value to the RAM buffer.
    * Note: You MUST call syncPNG(file, "commit") after calling this to save changes.
    */
    _putToPNG(key, value, keyPassword) {
        // Safety check: ensure buffer is initialized
        if (!this.dbBuffer || this.dbBuffer.length === 0) {
            this.dbBuffer = Buffer.alloc(48000000, 0);
        }
        const index = this._deriveIndex(key, keyPassword);
        const curD = this.hVal + (index * 1600);
        const { x, y } = this._d2xy(this.canvasSize, curD);
        // Linear pixel walk in RAM
        for (let i = 0; i < value.length; i++) {
            const pixelIdx = ((y * this.canvasSize) + (x + i)) * 3;
            if (pixelIdx + 2 < this.dbBuffer.length) {
                // Char code retrieval for bitwise XOR
                const charCode = value.charCodeAt(i);
                const secureByte = this._cryptByte(charCode, keyPassword, i);
                // XOR into Red Channel (Bit-perfect data preservation)
                this.dbBuffer[pixelIdx] = secureByte;
            }
        }
        // Write Null Terminator (0 in Red channel marks the end)
        const termIdx = ((y * this.canvasSize) + (x + value.length)) * 3;
        if (termIdx + 2 < this.dbBuffer.length) {
            this.dbBuffer[termIdx] = 0;
        }
    }

    /**
     * Retrieves a value from the RAM buffer.
     * Note: You SHOULD call _syncPNG(file, "load") before this to ensure fresh data.
     */
    _getFromPNG(key, keyPassword) {
        if (!this.dbBuffer || this.dbBuffer.length === 0) return "";
        const index = this._deriveIndex(key, keyPassword);
        const curD = this.hVal + (index * 1600);
        const { x, y } = this._d2xy(this.canvasSize, curD);
        let result = "";
        let i = 0;
        while (true) {
            const pixelIdx = ((y * this.canvasSize) + (x + i)) * 3;
            // Bounds check + Null Terminator check (Red == 0)
            if (pixelIdx + 2 >= this.dbBuffer.length || this.dbBuffer[pixelIdx] === 0) {
                break;
            }
            const scrambled = this.dbBuffer[pixelIdx];
            const decryptedByte = this._cryptByte(scrambled, keyPassword, i);
            result += String.fromCharCode(decryptedByte);
            i++;
        }
        return result;
    }

    // DB sync functions for easier DB handling
    _syncPNG(filename, mode) {
        mode = mode.toLowerCase();
        // Check if we are in LOAD MODE but the file is missing
        if (mode === "load" && !fs.existsSync(filename)) {
            console.log("⚠️ Database not found. Initializing new compressed store...");
            this.dbBuffer = Buffer.alloc(48000000, 0);
            return this._syncPNG(filename, "commit");
        }
        if (mode === "commit") { // COMMIT MODE (RAM -> Disk)
            const filtered = Buffer.alloc(48004000);
            for (let r = 0; r < 4000; r++) {
                filtered[r * 12001] = 0; // Filter byte
                this.dbBuffer.copy(filtered, r * 12001 + 1, r * 12000, (r + 1) * 12000);
            }
            const compressed = zlib.deflateSync(filtered, { level: 9 });
            try {
                const fd = fs.openSync(filename, 'w');   
                const sig = Buffer.from([0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A]);
                fs.writeSync(fd, sig);
                const ihdrData = Buffer.from([
                    0x49, 0x48, 0x44, 0x52, 
                    0x00, 0x00, 0x0F, 0xA0,
                    0x00, 0x00, 0x0F, 0xA0,
                    0x08, 0x02, 0x00, 0x00, 0x00 
                ]);
                fs.writeSync(fd, this._write32(13));
                fs.writeSync(fd, ihdrData);
                fs.writeSync(fd, this._write32(this._computeCRC(ihdrData, 17)));
                const idatHeader = Buffer.from("IDAT");
                fs.writeSync(fd, this._write32(compressed.length));
                fs.writeSync(fd, idatHeader);
                fs.writeSync(fd, compressed);
                const idatCrc = this._computeCRC(Buffer.concat([idatHeader, compressed]), compressed.length + 4);
                fs.writeSync(fd, this._write32(idatCrc));
                const iendData = Buffer.from("IEND");
                fs.writeSync(fd, this._write32(0));
                fs.writeSync(fd, iendData);
                fs.writeSync(fd, this._write32(this._computeCRC(iendData, 4)));
                fs.closeSync(fd);
                console.log(`✅ PNG Commit Successful: ${filename}`);
                return true;
            } catch (err) {
                return false;
            }

        } else if (mode === "load") { // LOAD MODE (Disk -> RAM)
            try {
                const file = fs.readFileSync(filename);
                const cLen = file.readUInt32BE(33);
                const cData = file.subarray(41, 41 + cLen);
                const decomp = zlib.inflateSync(cData);
                this.dbBuffer = Buffer.alloc(48000000);
                for (let r = 0; r < 4000; r++) {
                    decomp.copy(this.dbBuffer, r * 12000, r * 12001 + 1, (r + 1) * 12001);
                }
                //console.log(`✅ PNG Load Successful: ${filename}`);
                return true;
            } catch (err) {
                return false;
            }
        }
        return false;
    }

    _syncPPM(filename, mode) {
        mode = mode.toLowerCase();
        // Check if we are in LOAD MODE but the file is missing
        if (mode === "load" && !fs.existsSync(filename)) {
            console.log("⚠️ PPM Database not found. Initializing new raw store...");
            this.dbBuffer = Buffer.alloc(48000000, 0);
            return this._syncPPM(filename, "commit");
        }
        if (mode === "commit") { // COMMIT MODE (RAM -> Disk)
            try {
                const header = `P6\n4000 4000\n255\n`;
                const output = Buffer.concat([Buffer.from(header), this.dbBuffer]);
                fs.writeFileSync(filename, output);
                console.log(`✅ PPM Commit Successful: ${filename}`);
                return true;
            } catch (err) {
                return false;
            }
        } else if (mode === "load") { // LOAD MODE (Disk -> RAM)
            try {
                const file = fs.readFileSync(filename);
                // PPM header "P6\n4000 4000\n255\n" is exactly 15 bytes
                this.dbBuffer = file.subarray(15, 15 + 48000000);
                console.log(`✅ PPM Load Successful: ${filename}`);
                return true;
            } catch (err) {
                return false;
            }
        }
        return false;
    }

    // Blank DB creation functions
    _createPNG_DB(filename) {
        this.dbBuffer = Buffer.alloc(48000000, 0);
        if (this._syncPNG(filename, "commit")) {
            console.log(`✅ PNG initialized and loaded into buffer: ${filename}`);
        } else {
            console.error("❌ Failed to initialize PNG database file.");
        }
    }

    _createPPM_DB(dbPath) {
        try {
            const fd = fs.openSync(dbPath, 'w');
            // Write the P6 Header (Standard 4000x4000 8-bit RGB)
            const header = `P6\n${this.canvasSize} ${this.canvasSize}\n255\n`;
            fs.writeSync(fd, header);
            // Create one row of zeros (4000 pixels * 3 bytes)
            const zeroRow = Buffer.alloc(this.canvasSize * 3, 0);
            for (let i = 0; i < this.canvasSize; i++) {
                fs.writeSync(fd, zeroRow);
            }
            fs.closeSync(fd);
            console.log(`✅ PPM Database Ready (Isolated from RAM): ${dbPath}`);
        } catch (err) {
            console.error(`❌ Error: Could not create PPM database: ${err.message}`);
        }
    }

    // DB conversion functions
    _convertDatabaseFormat(pngPath, ppmPath, targetFormat) {
        // Standardize the flag to lowercase
        const format = targetFormat.toLowerCase();
        if (format === "ppm") {
            // SOURCE: PNG -> TARGET: PPM (Decompress and Expand)
            if (!this._syncPNG(pngPath, "load")) {
                console.error("❌ Failed to load PNG database.");
                return false;
            }
            try {
                // Create the PPM Header
                const header = Buffer.from("P6\n4000 4000\n255\n");
                // Concatenate header and the raw buffer
                const output = Buffer.concat([header, this.dbBuffer]);   
                fs.writeFileSync(ppmPath, output);
                console.log("✅ Converted PNG to PPM (48MB Raw Volume)");
                return true;
            } catch (err) {
                return false;
            }
        } else if (format === "png") {
            // SOURCE: PPM -> TARGET: PNG (Pack and Compress)
            try {
                const fullFile = fs.readFileSync(ppmPath);
                // Skip the header (15 bytes) and extract the raw 48MB
                // subarray() provides a view without copying the memory
                const rawDataSize = 48000000;
                this.dbBuffer = fullFile.subarray(15, 15 + rawDataSize);
                // Use syncPNG to compress and save as a PNG
                if (!this._syncPNG(pngPath, "commit")) {
                    console.error("❌ Failed to compress and save PNG database.");
                    return false;
                }
                console.log("✅ Converted PPM to PNG");
                return true;
            } catch (err) {
                return false;
            }
        } else {
            console.error("❌ Invalid target format. Use 'PNG' or 'PPM'.");
            return false;
        }
    }

    // DB Deletion and cleanup functions
    _deleteDB(dbPath) {
        try {
            fs.unlinkSync(dbPath); // Synchronous, no callback needed
            return true;
        } catch (error) {
            if (error.code === 'ENOENT') return false;
            throw error;
        }
    }

    // DB checking/validating functions
    _getFileSize(path) {
        try {
            // Check if path exists and is a regular file
            const stats = fs.statSync(path);
            if (stats.isFile()) {
                return stats.size;
            }
        } catch (err) {
            // console.error(`❌ Filesystem Error: ${err.message}`);
        }
        return 0;
    }

    _validateDB(path) {
        // 4000 * 4000 * 3
        const rawDataSize = 48000000;
        const actual = this._getFileSize(path);
        if (path.includes(".ppm")) {
            // PPM must have the header + the data (Header is ~15 bytes)
            return actual >= (rawDataSize + 15);
        } else {
            // PNG/Raw must be at least the data size
            return actual >= rawDataSize;
        }
    }

}

module.exports = ScyKernel;